### ISSUE TITLE
Add more detail to student on course instance - progression codes, "outcomes"

### ISSUE THREAD

**Issue by arc12** [jiscdev/analytics-udd#132]
Idea for 1.3

SITS (and I assume other systems, because these related to key aspects of academic process) has information on the student's status at the end of a course instance which, among other things, indicates their eligiblity to progress.

Some examples:
  "FR" : "Fail - Repeat failed modules in next academic year",
  "FS" : "Fail - Restart whole stage in next academic year",
  "JE" : "In Jeopardy - withdraw from the programme",
  "PH" : "Progress returning to Honours",
  "PO" : "Progress but transfer to Ordinary degree programme",
  "PP" : "Progress - Proceed to next stage",
  "PT" : "Progress - Proceed with trailing modules",
  "REF" : "Referral",
  "SC" : "Successful Completion",
  "WD" : "Fail - Withdraw",
  "WN" : "Withdraw - for non academic reasons"

There is clearly some overlap but there is also additional info which can't easily (if at all) be inferred from current UDD fields.

**Comment by alanepaull** [comment]
After discussions with colleagues on Wednesday (and per @arc12 initial comment), I think that this issue affects student_on_course_instance, and only indirectly student_course_membership.  At the moment all the progression information related to the student on the course is held in properties of student_course_membership, for example WITHDRAWAL_REASON, COURSE_OUTCOME.  Student_on_course_instance contains no progression information.

My understanding is that course_instance relates to a particular academic year of a course.  The implication here is that students progress from one course_instance to another, so it seems likely that progression information is very relevant to student_on_course_instance.

Clearly, progression information at course level would usually refer to the student's completion / outcome, or, prior to completion, the student's continuation on the course.  It's important, I think, not to muddle course_instance with course; for example, I've just noticed that the COURSE_GRADE definition on COURSE refers to achievement on this "course_instance"; so, it's easy to do.  Potential confusion may be caused by the name "course_instance", which sounds like a concrete example of a whole course, rather than a stage.

I agree with @arc12 that we probably need a generic set of descriptions and codes for a student's progression at the end of a course_instance.  Some of these descriptions may look similar to course outcomes and may even overlap with them.

Is there a wider issue that course_instance may not solely relate to an academic year?  Some courses (PG for instance) may have progression points after 6 months.  Would these need to be covered more explicitly than the current UDD properties?

**Comment by alanepaull** [comment]
From Monday 5 Dec call:

Source will be more detailed than proposed UDD vocabulary, and institution-specific, so institutions may want to dig into their own stuff.

- Introduce a new property to hold source data
- Create generic, probably sparse mapped vocabulary for UDD.

**Comment by alanepaull** [comment]
HESA has the CSTAT field that offers a generic vocab for this, so I've used that as the basis for a new progression property, with progression_source to hold the source data.
For discussion.

**Comment by arc12** [comment]
Alan -
I think that is OK, although we should check that the mapping of local progression codes to CSTAT is going to be easy-ish in the UDD ETL.

Given that FE doesn't (AFAIK) have the same notion of a "course instance", although we'll still probably want to use the same UDD entities, is the implication that these fields are not applicable to FE. i.e that fixing on CSTAT is not an issue for cross-sector use of UDD?

On a point of documentation:
we have a python script (in this repo - https://github.com/jiscdev/analytics-udd/blob/v1.3/utilities/Extract%20Code%20Lists%20from%20MD.py) which extracts code lists to JSON from the markdown. This assumes the code list table header contains the attribute name, to help in matching...
It would be helpful if new tables followed the convention (otherwise whoever ends up generating the JSON code lists will end up issuing a pull request with new table headings)

**Comment by alanepaull** [comment]
@arc12: I would be interested to see how course_instance is represented in SITS.  Do you have any examples?  I made a statement above, "My understanding is that course_instance relates to a particular academic year of a course. The implication here is that students progress from one course_instance to another, so it seems likely that progression information is very relevant to student_on_course_instance."  I now think that this is substantially incorrect - it seems that I may not quite have understood course_instance properly, caused partly by entering the project relatively recently.  However, I'd like to check this out.

Now that I've seen some data, it's clear that University of Jisc data has a course_instance as a whole course, for example a 3 yr Hons Degree starting in 2014, with a start date in 2014 and end date in 2017.  This would mean that course_instance is pretty much equivalent to xcri-cap presentation (though 'running' rather than 'offered').  This would match with HESA data.

My slight concern is your earlier statement (at the top of this issue): "...has information on the student's status at the end of a course instance which, among other things, indicates their eligibility to progress."  The end of a course_instance, if it's a whole course, would be a final outcome, not a stage within a course from which they progress to the next stage.

I hope you can shed some light on this for me.

**Comment by arc12** [comment]
@alanepaull - Craig Petch is your man for that question.
The first(ish) iteration of UDD did not have Student Course Membership, and one of the earliest points of discussion (before use of github) was around this point. I recall coming along with the xcri concept and being conceptually recalibrated! It may be that the UoJ data is obsolete.
My understanding is that Quercus has the unit of progression concept too. Indeed, it is clear that we do need to track annual progression points for HE, and that YEAR_PRG in SoCI is predicated on CI being an annual cycle, otherwise SoCI would need to include the academic year info in its uniqueness constraint.

So...
1) I think the unit of progression concept is intended
2) this undelines the importance of documenting the concept/function of entities from the start!

**Comment by craig-petch** [comment]
Alan. The original concept was that a course instance  would capture each enrolment instance of a course which is typically an academic year, this also normally relates to a progression period of some description, or a least to a summary of a student's progression status at the end of that instance.  I'm not sure if this helps, but looking at [HEDIIP Data Language dictionary](https://hediip.ac.uk/wp-content/uploads/Data-Dictionary-2016-07-28.pdf) this is similar in concept to the Course Session Entity (pg. 38), although I don't think we are anticipating getting into some potential progression complexity.  If it would help to look at some SITS based example data let me know and we can arrange a web meeting and work through some scenarios

**Comment by alanepaull** [comment]
That does make sense - I can use that as the basis for our description / definition of course_instance.  The main concern would be that everyone approaches it in the same way, I guess.

It might be a good idea to see some data, if we can arrange a suitable time.  Any suggestions?

Alan

**Comment by craig-petch** [comment]
How about Friday 16th any time between 2-4 pm or Monday 19th 11-12 am or 2-3 pm ?

**Comment by alanepaull** [comment]
Friday's difficult.  Monday at 11am would be great.  My Skype is alanepaull.
Thanks, Craig.
Alan

**Comment by alanepaull** [comment]
Progression codes have been updated with some more detail added, in line with our discussions.  I have attempted to make these as generic as possible, while retaining derivation from HESA CSTAT.  While the value format is String, I've made the codes 2 digits with no leading zeros, so it is possible to insert additional codes without too much problem.  I've not created abbreviated alphabetical codes, because there is a danger of mismatches between UDD codes and proprietary codes, which might lead to confusion.
Alan

**Comment by arc12** [comment]
Query:
These two seem very similar:
15 	N/A 	The student is continuing on the course and is repeating some modules from this stage alongside the next stage modules.

50 	N/A 	The student has failed this stage of the course and is repeating some modules, but not the whole stage.

Should "15" be explicit that the student has not failed the stage?
Should "50" be explicit that they are doing the repeats alongside the next stage? Alternatively that they are not doing them alongside the next stage.

In other words: in what ways do these two categories differ?

**Comment by alanepaull** [comment]
15 = not failed; does that need to be explicit or is "continuing" sufficient?
50 = The student has failed this stage of the course and is repeating some modules, but not the whole stage, and is not yet taking the next stage.

Do we need 50?  It seems esoteric, but I guess it could be a student re-taking modules over the summer?  Or over the semester or academic year, effectively as an extra year?  I think there will be some institutions with this possibility.

Revision:

15 N/A The student is continuing on the course and is repeating some modules from this stage alongside the next stage modules; the student has not failed the stage.

50 N/A The student has failed this stage of the course and is repeating some modules, but not the whole stage; the student is not yet continuing to the next stage.

**Comment by arc12** [comment]
I'm not really in a position to comment on "do we need?"
My comment was purely a reflection of trying to internalise the categories.

Craig may have a better picture.
I think this is something to determine based on what progression codes are in use across a sample of institutions. I've seen the Aston codes, but not the explanation of what they mean, which I think is necessary to decide: could the institutionally-specific codes be mapped without having to make too many arbitrary decisions about box A vs box B because our boxes segmented things across actual categories in use.

**Comment by arc12** [comment]
I move to include in v1.3.0

I think we are close to agreement.

**Comment by alanepaull** [comment]
Monday 13 February: Agreed to modify PROGRESSION as suggested in alanepaull above; may be subject to modification later in the light of experience.  PROGRESSION_SOURCE is useful immediately, in any case.
